src/daft-logical-plan/src/optimization/rules/unnest_subquery.rs

```diff
@@ -0,0 +1,779 @@
+
+        let new_input = subqueries
+            .into_iter()
+            .try_fold(input, |curr_input, subquery| {
+                let subquery_alias = subquery.semantic_id().id;
+                let subquery_plan = downcast_subquery(&subquery);
+
+                let subquery_col_names = subquery_plan.schema().names();
+                let [output_col] = subquery_col_names.as_slice() else {
+                    return Err(DaftError::ValueError(format!(
+                        "Expected scalar subquery to have one output column, received: {}",
+                        subquery_col_names.len()
+                    )));
+                };
+
+                // alias output column
+                let subquery_plan = Arc::new(LogicalPlan::Project(Project::try_new(
+                    subquery_plan,
+                    vec![col(output_col.as_str()).alias(subquery_alias)],
+                )?));
+
+                let (decorrelated_subquery, subquery_on, input_on) =
+                    pull_up_correlated_cols(subquery_plan)?;
+
+                if subquery_on.is_empty() {
+                    // uncorrelated scalar subquery
+                    Ok(Arc::new(LogicalPlan::Join(Join::try_new(
+                        curr_input,
+                        decorrelated_subquery,
+                        vec![],
+                        vec![],
+                        None,
+                        JoinType::Inner,
+                        None,
+                        None,
+                        None,
+                        false,
+                    )?)))
+                } else {
+                    // correlated scalar subquery
+                    Ok(Arc::new(LogicalPlan::Join(Join::try_new(
+                        curr_input,
+                        decorrelated_subquery,
+                        input_on,
+                        subquery_on,
+                        None,
+                        JoinType::Left,
+                        None,
+                        None,
+                        None,
+                        false,
+                    )?)))
+                }
+            })?;
+
+        Ok(Transformed::yes((new_input, new_exprs)))
+    }
+}
+
+impl OptimizerRule for UnnestScalarSubquery {
+    fn try_optimize(&self, plan: Arc<LogicalPlan>) -> DaftResult<Transformed<Arc<LogicalPlan>>> {
+        plan.transform_down(|node| match node.as_ref() {
+            LogicalPlan::Filter(Filter {
+                input, predicate, ..
+            }) => {
+                let unnest_result =
+                    Self::unnest_subqueries(input.clone(), split_conjuction(predicate))?;
+
+                if !unnest_result.transformed {
+                    return Ok(Transformed::no(node));
+                }
+
+                let (new_input, new_predicates) = unnest_result.data;
+
+                let new_predicate = conjuct(new_predicates)
+                    .expect("predicates should exist in unnested subquery filter");
```

Comment:
nit: this implies that there's a potential for it to not exist. I'd change the wording slightly
```suggestion
                let new_predicate = conjuct(new_predicates)
                    .expect("predicates are guaranteed to exist at this point, so 'conjunct' should never return 'None'");
```